### PR TITLE
Use a memory backend for the upper layers

### PR DESCRIPTION
**Pull request by icristescu**
I switched the upper layer to use `pack_mem` which is basically `irmin_mem` (but I had issues with the functors exposed by `irmin_mem`, we can clean it up later). For now most tests fails.

### CODE REVIEW

**Comment by icristescu** [comment]
There are still a few tests missing (watches, integration and migration tests), but it could use some early reviews. 

Based on https://github.com/mirage/irmin/pull/1111.

**Comment by icristescu** [comment]
Rebased and ready for reviews. 

In particular, I hesitate in keeping this [extra check](https://github.com/mirage/irmin/pull/1099/files#diff-0656cbce221ba6715aa8e99bc19aea4bd241093462b5c68f4671223326349928R757-R779). It is useful if we use a lower layer and by mistake we attempt to copy objects from it back to upper (for instance when the upper is not self contained), but it is a possibly costly extra check otherwise.

File: src/irmin-test/store.ml [hunk 1]
```diff
@@ -772,7 +772,7 @@ module Make (S : S) = struct
             watch (n - 1)
       in
       let v1 = "X1" in
-      let v2 = "X2" in
+      let _v2 = "X2" in
```
Comment (CraigFe) [inline src/irmin-test/store.ml:775]: This one's a bit odd; shouldn't it have been a warning if it wasn't being used? (In which case, we should just be able to remove it outright.)

**Review by CraigFe — APPROVED** [review]
LGTM. I'm a bit sad to see so many extra functors, but we can fix that in a later iteration.

File: src/irmin-test/store.ml (hunk 1 above)
Comment (icristescu) [reply src/irmin-test/store.ml]: Good catch, thank you!

**Comment by icristescu** [comment]
> LGTM. I'm a bit sad to see so many extra functors, but we can fix that in a later iteration.

I was hoping you can make suggestions on that :) but it's probably easier to do in a separate PR.

**Comment by icristescu** [comment]
Closing this as we probably should do this in two steps:
- add an irmin_pack.mem backend (so bits of this PR are now copied in https://github.com/mirage/irmin/pull/1280);
- then add the irmin_pack.mem backend as a layer in the layered store.

**Pull request closed without merging**